muffnn/autoencoder/tests/test_autoencoder.py

```diff
@@ -0,0 +1,539 @@
+"""
+Tests for the Autoencoder.
+"""
+
+import logging
+import pprint
+from io import BytesIO
+import pickle
+
+import pytest
+import numpy as np
+import scipy.special
+import scipy.sparse as sp
+import tensorflow as tf
+from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
+from sklearn.datasets import load_iris
+from sklearn.utils.estimator_checks import check_estimator
+from sklearn.utils.testing import (assert_array_almost_equal,
+                                   assert_almost_equal)
+
+from muffnn import Autoencoder
+
+
+_LOGGER = logging.getLogger(__name__)
+iris = load_iris()
+
+
+def sigmoid_cross_entropy(ytrue, ypred):
+    """Compute the cross-entropy for a set of [0, 1] labels and output
+    probabilities for each label.
+
+    This function computes
+
+        - ytrue * np.log(ypred) - (1 - ytrue) * np.log(1 - ypred)
+
+    This function uses the numerically stable version from TensorFlow
+
+        np.max(x, 0) - x * z + np.log(1 + np.exp(-np.abs(x)))
+
+    where `x` is the logit of `ypred` and `z` is `ytrue`.
+
+    It also handles 0's and 1's specially since they can produce NaNs with
+    the TensorFlow version.
+    """
+
+    ce = np.zeros_like(ytrue)
+    x = scipy.special.logit(ypred)
+    z = ytrue
+    q = np.isfinite(x)
+    ce[q] = (np.clip(x[q], 0.0, np.inf) - x[q] * z[q] +
+             np.log(1.0 + np.exp(-1.0 * np.abs(x[q]))))
+
+    # handle 1's and 0's specially
+    q = (((ypred == 1.0) & (ytrue == 1.0)) |
+         ((ypred == 0.0) & (ytrue == 0.0)))
+    ce[q] = 0.0
+
+    q = (((ypred == 1.0) & (ytrue != 1.0)) |
+         ((ypred == 0.0) & (ytrue != 0.0)))
+    ce[q] = np.inf
+
+    return np.sum(ce, axis=1)
+
+
+def softmax_cross_entropy(ytrue, ypred):
+    """Compute the cross-entropy for a multi-class classifier.
+
+    This function computes
+
+        np.sum(-ytrue * np.log(ypred), axis=1)
+
+    It also handles 0's since they can produce NaNs.
+
+    See scikit.metrics.log_loss for a version with clipping of the
```

Comment:
I remember why now. The sklearn one sums over the samples auotmagically. I wanted to test the score of each sample.